### PR TITLE
[build support] Generic prefixed binary lookup function in swift_build_support

### PR DESCRIPTION
**Pull request by harlanhaskins**
#### What's in this pull request?

A small modification to `swift_build_support` to allow looking up arbitrary commands like `llvm-cov` and `llvm-profdata` alongside `clang` and `clang++`.
#### Resolved bug number: N/A

Will help with [SR-237](https://bugs.swift.org/browse/SR-237).

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by harlanhaskins** [comment]
@modocache @gribozavr Thoughts?

**Comment by harlanhaskins** [comment]
Actually, this doesn't take the native llvm/clang tools option into account, regardless. That needs to be addressed.

**Comment by harlanhaskins** [comment]
Any ideas here? The current functions don't take into account `NATIVE_[LLVM|CLANG]_TOOLS_PATH`, which is pretty important.

**Comment by harlanhaskins** [comment]
@gribozavr How about this?

File: utils/swift_build_support/swift_build_support/toolchain.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+from .which import which
+
+
+class Toolchain(object):
+    """
+    A class that stores accessible paths to system commands.
+
+    It requires a 'cc' and 'cxx' tool in its keyword args.
+    """
+    def __init__(self, cc=None, cxx=None, **kwargs):
```
Comment (modocache) [inline utils/swift_build_support/swift_build_support/toolchain.py]: Sorry if I'm not picking up on something here, but why provide default arguments for arguments the documentation claims are necessary? If you're looking for the ability to use named arguments like `Toolchain(cc="foo", cxx="bar")`, you can do that without supplying default arguments in Python.

File: utils/swift_build_support/swift_build_support/toolchain.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        suffixes = ['']
+    for suffix in suffixes:
+        path_map = dict()
+        for name, tool in tools.iteritems():
+            path = which(tool + suffix)
+            if not path:
+                break
+            path_map[name] = path
+        if len(path_map) == len(tools):
+            return Toolchain(**path_map)
```
Comment (modocache) [inline utils/swift_build_support/swift_build_support/toolchain.py]: Oh, is this why default arguments are provided above? A bit too much metaprogramming for me, but to each their own! :sweat_smile:

File: utils/swift_build_support/swift_build_support/toolchain.py (hunk 1 above)
Comment (harlanhaskins) [inline utils/swift_build_support/swift_build_support/toolchain.py]: I figured it was good to express that these arguments are strictly required, though the same could be expressed with

`assert(tools.get('cc') is not None)`

File: utils/swift_build_support/tests/test_toolchain.py [hunk 3]
```diff
@@ -26,6 +27,18 @@ def test_clang_available_on_this_platform(self):
         self.assertTrue(os.path.split(clang.cc)[-1].startswith('clang'))
         self.assertTrue(os.path.split(clang.cxx)[-1].startswith('clang++'))
 
+    def test_found_executables_match(self):
```
Comment (modocache) [inline utils/swift_build_support/tests/test_toolchain.py]: nit-pick: This doesn't test `host_clang()`, it tests `host_toolchain()`. As such, I'd recommend putting it in a separate `class HostToolchainTestCase(unittest.TestCase)`.

**Comment by modocache** [comment]
The commit message claims these changes will allow searching for arbitrary executables besides `clang`, like `llvm-cov`. Can you add a test case to prove it?

**Comment by harlanhaskins** [comment]
I can add a test case, but I don't want tests to fail if the tester actually doesn't have the tools installed...

**Comment by modocache** [comment]
Can we assume that certain tools will always exist on OS X systems? If so you can run them on OS X only, like these are: https://github.com/apple/swift/blob/master/utils/swift_build_support/tests/test_xcrun.py#L18-L20

**Comment by harlanhaskins** [comment]
That's totally doable. IIRC, Xcode includes, at the very least, `llvm-profdata`, right @gribozavr?

**Comment by modocache** [comment]
Sure! I also don't think the test has to be for something related to profdata, just something besides Clang.

**Comment by harlanhaskins** [comment]
@modocache Thoughts on the extra test?

**Comment by modocache** [comment]
Looks good to me! @gribozavr could you ask @swift-ci to please test?

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by harlanhaskins** [comment]
Tests pass. @gribozavr Good to merge?

**Comment by shahmishal** [comment]
@swift-ci Please python lint

**Comment by shahmishal** [comment]
```
--- Python lint ---
./utils/swift-api-dump.py:98:80: W291 trailing whitespace
./utils/swift-api-dump.py:99:80: E501 line too long (82 > 79 characters)
./utils/swift_build_support/swift_build_support/toolchain.py:2:80: E501 line too long (80 > 79 characters)
./utils/swift_build_support/swift_build_support/toolchain.py:21:1: F401 'collections' imported but unused
./utils/swift_build_support/tests/test_toolchain.py:14:1: I100 Import statements are in the wrong order. import platform
```

**Comment by shahmishal** [comment]
@swift-ci Please python lint

**Comment by shahmishal** [comment]
@harlanhaskins Thanks for fixing the python lint errors.

**Comment by harlanhaskins** [comment]
@gribozavr @modocache Think this is good to merge?

**Comment by gribozavr** [comment]
LGTM.

@swift-ci Please test

**Comment by harlanhaskins** [comment]
That test failure is unrelated...

**Comment by Geforce8472** [comment]
This change seems to leave me unable to build swift on Fedora Linux. 

which: no clang-3.8 in (/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin)
which: no clang-3.7 in (/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin)
which: no clang-3.6 in (/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin)
which: no clang-3.5 in (/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin)

clang exists in /usr/bin currently version 3.7

File: utils/swift_build_support/swift_build_support/toolchain.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            # On newer releases of FreeBSD, the default Clang is sufficient.
+            return host_toolchain(xcrun_toolchain)
+        else:
+            # On older releases, or on releases for which we cannot determine
+            # the release date, we search for the most modern version
+            # available.
+            return host_toolchain(xcrun_toolchain,
+                                  suffixes=['38', '37', '36', '35'])
+    return host_toolchain(xcrun_toolchain,
+                          suffixes=['-3.8', '-3.7', '-3.6', '-3.5'])
```
Comment (rintaro) [inline utils/swift_build_support/swift_build_support/toolchain.py]: `suffixes` should contain `'', ` [as before](https://github.com/apple/swift/pull/1586/files#diff-0084283925791f53be5c74fd2938d392L90)

Comment (harlanhaskins) [inline utils/swift_build_support/swift_build_support/toolchain.py]: Yep, you're right. Thanks for catching that! Fixed in master.

**Comment by RLovelett** [comment]
Unfortunately this commit broke Python 3 support.

```
$ utils/build-script --release -- --reconfigure
Traceback (most recent call last):
  File "utils/build-script", line 1167, in <module>
    sys.exit(main())
  File "utils/build-script", line 1163, in main
    return main_normal()
  File "utils/build-script", line 1023, in main_normal
    xcrun_toolchain=args.darwin_xcrun_toolchain)
  File "utils/swift_build_support/swift_build_support/toolchain.py", line 124, in host_clang
    suffixes=['-3.8', '-3.7', '-3.6', '-3.5'])
  File "utils/swift_build_support/swift_build_support/toolchain.py", line 101, in host_toolchain
    return _first_common_toolchain(tools, suffixes=suffixes)
  File "utils/swift_build_support/swift_build_support/toolchain.py", line 70, in _first_common_toolchain
    for name, tool in tools.iteritems():
AttributeError: 'dict' object has no attribute 'iteritems'
```

**Pull request merged**